# Processo Seletivo Estágio Syngenta

## Tarefa 1 - Conte a quantidade de pixels verdes 

Para esta primeira tarefa, usaremos a biblioteca opencv para manipularmos a imagem .bmp fornecida a fim de obter as informações que desejamos.
Abaixo importamos os módulos que usaremos para a tarefa e carregamos a imagem para manipularmos esta.

In [1]:
import cv2 as cv

img = cv.imread('resources/Syngenta.bmp')
img


array([[[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       ...,

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
        ...,
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]]], dtype=uint8)

A imagem é carregada como um Numpy array, no qual cada pixel é formado por um outro array de 3 valores, adotando a notação RGB para definir qual a cor do pixel em questão (Nota: a biblioteca openCV adota a notação dos valores como BGR e não RGB, como é usual).

A próxima etapa agora é identificar quais cores estão presentes na imagem.

In [2]:
import numpy as np

img_flat = img.reshape(-1, img.shape[2])
unicas = np.unique(img_flat, axis=0)
unicas

array([[  0,   0,   0],
       [  0, 192,  96],
       [255, 255, 255]], dtype=uint8)

Aqui nós primeiro "achatamos" o array, de forma que o total de linhas do array seja igual ao total de pixels da imagem. Depois buscamos os elementos únicos que compõem o array, especificando o eixo das linhas (axis = 0), para considerarmos a tupla BGR ao invés de cada valor de forma individual.

Identificamos que há 3 cores presentes na imagem. Dada a notação RGB para a cor de cada pixel, sabemos que a tupla (0, 0, 0) representa a cor preta e que a tupla (255, 255, 255) representa a cor branca. Portanto, isso nos leva a concluir que a tupla (0, 192, 96) representa a cor verde que buscamos.

Agora, nos resta contar quantos pixels da imagem são representados pela tupla (0, 192, 96) e teremos, assim, o total de pixels verdes na imagem. Para tanto, vale notar que os valores RGB que diferem a cor verde da preta e da branca são relativos às camadas R e G, logo, podemos realizar um slice no array da imagem a fim de selecionarmos apenas uma dessas camadas e assim facilitar o processo computacional de contagem.

In [3]:
img_slice = img[:,:, 2]
mascara = (img_slice == unicas[1][2])
total_verdes = np.count_nonzero(mascara)
total_verdes

298

Assim, selecionamos apenas o canal R do array da imagem e aplicamos uma máscara neste, de forma que obtemos um array booleano composto por majoritariamente valores False (0) e valores True (1) apenas para as posições dos pixels cujos valores sejam de 96, o qual é o valor do canal R que compõe a cor verde presente na imagem, como identificamos acima.

Portanto, temos um array onde temos o valor 1 para cada pixel verde presente na imagem, podemos então realizar a soma dos valores não nulos presentes neste array e teremos como resultado desta soma o número total de pixels verdes presentes na imagem, o qual é de 298 pixels verdes.

## Tarefa 2 - A imagem possui uma mensagem escondida, descubra qual a mensagem

### Tentativa 1 - Stenografia

A stenografia consiste em esconder informações secretas em locais não secretos